### PR TITLE
Allow to call mDNS browse callback more than once

### PR DESCRIPTION
**Pull request by arkq**
#### Problem

Currently it's not possible to call DNS-SD browse callback more than once (proxy object acquired in `FindCommissionableNodes` or `FindCommissioners` is released in this callback). However, generic OpenThread implementation [calls this callback in a loop](https://github.com/project-chip/connectedhomeip/blob/master/src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp#L2577).

#### Change overview

This commit extends mDNS browsing callback to be called once per single mDNS service or all at once. In case when the callback might be called again, the internal proxy is not released.

#### Testing

For now the overall behavior should not be changed. CI should test that.

### CODE REVIEW

**Review by bzbarsky-apple — COMMENTED** [review]
This generally makes sense, but it seems like we should also have an API for _stopping_ a browse, as a consumer, if we're going to do this.

@vivien-apple

File: src/lib/dnssd/platform/Dnssd.h [hunk 1]
```diff
@@ -209,6 +214,27 @@ CHIP_ERROR ChipDnssdFinalizeServiceUpdate();
...
+ * @param[in] addressType The protocol version of the IP address.
+ * @param[in] interface   The interface on which queries are sent.
+ *
+ * @retval CHIP_NO_ERROR                The stop browse succeeds.
+ * @retval Error code                   The stop browse fails.
+ *
+ */
+CHIP_ERROR ChipDnssdStopBrowse(const char * type, DnssdServiceProtocol protocol, chip::Inet::IPAddressType addressType,
+                               chip::Inet::InterfaceId interface);
+
```
Comment (vivien-apple) [inline src/lib/dnssd/platform/Dnssd.h:237]: A method to stop browsing looks nice to me. That said in the current shape I'm sort of afraid that we it is easy to leak a `ResolverProxy`. Looking into the Tizen implementation it seems like the code does remove the matching contexts from the internal list of context but nobody will ever call `proxy->release()` on them.

Should we call the default callback for those contexts, passing a service size of `0` and with `CHIP_ERROR_END_OF_INPUT` ?

If no browse operation is in progress I assume that `CHIP_NO_ERROR` is returned too ? Can we add that to the comment ?

File: src/platform/tests/TestDnssd.cpp [hunk 2]
```diff
@@ -14,39 +14,56 @@ using chip::Dnssd::DnssdService;
...
+    unsigned int n      = gResolvedServicesCount++;
 
     NL_TEST_ASSERT(suite, result != nullptr);
     NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
+
     if (!addresses.empty())
     {
         addresses.data()[0].ToString(addrBuf, sizeof(addrBuf));
-        printf("Service at [%s]:%u\n", addrBuf, result->mPort);
+        printf("Service[%u] at [%s]:%u\n", n, addrBuf, result->mPort);
```
Comment (vivien-apple) [inline src/platform/tests/TestDnssd.cpp:34]: Can't we just use `gResolvedServicesCount` instead of `n` ?

File: src/platform/tests/TestDnssd.cpp [hunk 3]
```diff
@@ -14,39 +14,56 @@ using chip::Dnssd::DnssdService;
...
     nlTestSuite * suite = static_cast<nlTestSuite *>(context);
 
-    NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
-    if (services)
+    // Make sure that we will not be called again after end-of-input is set
+    NL_TEST_ASSERT(suite, gEndOfInput == false);
+    NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR || error == CHIP_END_OF_INPUT);
+
+    gBrowsedServicesCount += servicesSize;
+    gEndOfInput = error == CHIP_END_OF_INPUT;
```
Comment (vivien-apple) [inline src/platform/tests/TestDnssd.cpp:58]: nit: `gEndOfInput = (error == CHIP_END_OF_INPUT);`

File: src/platform/tests/TestDnssd.cpp [hunk 4]
```diff
@@ -14,39 +14,56 @@ using chip::Dnssd::DnssdService;
...
     {
-        printf("Mdns service size %u\n", static_cast<unsigned int>(servicesSize));
-        printf("Service name %s\n", services->mName);
-        printf("Service type %s\n", services->mType);
-        NL_TEST_ASSERT(suite, ChipDnssdResolve(services, chip::Inet::InterfaceId::Null(), HandleResolve, suite) == CHIP_NO_ERROR);
+        printf("Browse mDNS service size %u\n", static_cast<unsigned int>(servicesSize));
+        for (size_t i = 0; i < servicesSize; i++)
+        {
+            NL_TEST_ASSERT(suite, ChipDnssdResolve(&services[i], services[i].mInterface, HandleResolve, suite) == CHIP_NO_ERROR);
+        }
```
Comment (vivien-apple) [inline src/platform/tests/TestDnssd.cpp:66]: Can we keep the logs ?

File: src/lib/dnssd/platform/Dnssd.h (hunk 1 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h:237]: >  Looking into the Tizen implementation it seems like the code does remove the matching contexts from the internal list of context but nobody will ever call proxy->release() on them.

Yes, that might be true, I will have to rethink that.

Also, for completeness, I though about adding browsing timeout to `ChipDnssdBrowse()`. I think it could be nice to be able to setup browsing for let say 15 seconds, then automatically stop it without the need of calling `ChipDnssdStopBrowse()`. Internally platform could implement it with a timer and call to `ChipDnssdStopBrowse`. What do you think about such addition?

Comment (vivien-apple) [reply src/lib/dnssd/platform/Dnssd.h:237]: > Also, for completeness, I though about adding browsing timeout to ChipDnssdBrowse(). I think it could be nice to be able to setup browsing after let say 15 seconds, then automatically stop it without the need of calling ChipDnssdStopBrowse(). Internally platform could implement it with a timer and call to ChipDnssdStopBrowse. What do you think about such addition?

That sounds like a followup. I would have argued that sometimes someone may want to do long lived browse queries and may not desire this behaviour so it should be configurable at the API level.
My understanding is that most platforms at the moment are sort of terminating the browse instructions early, but it is unclear to me that the needs for long lived browse requests may not appears one day.

Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h:237]: >  I would have argued that sometimes someone may want to do long lived browse queries and may not desire this behaviour so it should be configurable at the API level.

Yes, that what I had in mind. Something like this (with proper type of course):
```
CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, chip::Inet::IPAddressType addressType,
                           chip::Inet::InterfaceId interface, DnssdBrowseCallback callback, void * context, duration timeout = kDurationInfinite);
```

File: src/platform/tests/TestDnssd.cpp (hunk 4 above)
Comment (Arkq) [reply src/platform/tests/TestDnssd.cpp:67]: OK, I'm keeping the logs.

File: src/lib/dnssd/platform/Dnssd.h (hunk 1 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h:242]: > Looking into the Tizen implementation it seems like the code does remove the matching contexts from the internal list of context but nobody will ever call proxy->release() on them.
> Should we call the default callback for those contexts, passing a service size of 0 and with CHIP_ERROR_END_OF_INPUT ?

For Tizen impl I've solved this problem exactly like that. Also, I've added a note to the `ChipDnssdStopBrowse` docstring:

```
Please note, that when stopping mDNS browsing, it is REQUIRED to call the callback
passed to the `ChipDnssdBrowse` function with an error code or the CHIP_END_OF_INPUT
code. This will allow the mDNS module to release its resources.
```

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 5]
```diff
@@ -2481,6 +2481,7 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnDnsBrowseResult(otEr
             // Invoke callback for every service one by one instead of for the whole list due to large memory size needed to
             // allocate on
             // stack.
+            // FIXME (a.bokowy): Could we call callback multiple number of times (before this patch)???
```
[the remark this answers is not in the document]
Comment (Arkq) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp]: I think that I've fixed the issue here (see this commit: https://github.com/project-chip/connectedhomeip/pull/19554/commits/8e24dcfedddf9137b97c37e92285ffef4477b0b7), but this shall be tested. Unfortunately, right now, I've got no Thread-capable device at hand... I will be able to test that in few days. Alternatively someone with such devices (I think that one would need at least 3 devices) can test following scenario:

- setup two Thread-enabled devices with the same mDNS service
- run 3rd device in a browsing mode and see whether current master branch will report double-free and if this PR fixes that

File: src/lib/dnssd/platform/Dnssd.h [hunk 6]
```diff
@@ -102,10 +102,15 @@ using DnssdResolveCallback = void (*)(void * context, DnssdService * result, con
  * The callback function SHALL NOT take the ownership of the service pointer or
  * any pointer inside this structure.
  *
+ * It is possible that this callback will be called multiple number of times.
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/platform/Dnssd.h:105]:
```suggestion
 * It is possible that this callback will be called multiple times.
```

File: src/lib/dnssd/platform/Dnssd.h [hunk 7]
```diff
@@ -205,6 +210,32 @@ CHIP_ERROR ChipDnssdFinalizeServiceUpdate();
...
+ * @param[in] protocol    The service protocol.
+ * @param[in] addressType The protocol version of the IP address.
+ * @param[in] interface   The interface on which queries are sent.
+ *
+ * @retval CHIP_NO_ERROR                The stop browse succeeds or no browse for
+ *                                      given parameters is in progress.
+ * @retval Error code                   The stop browse fails.
+ *
+ */
+CHIP_ERROR ChipDnssdStopBrowse(const char * type, DnssdServiceProtocol protocol, chip::Inet::IPAddressType addressType,
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/platform/Dnssd.h:236]: Is there a public API that actually ends up calling this?  Or is that planned to be a followup?

File: src/platform/ESP32/DnssdImpl.cpp [hunk 8]
```diff
@@ -311,6 +311,8 @@ CHIP_ERROR OnBrowseDone(BrowseContext * ctx)
                 currentResult = currentResult->next;
                 servicesIndex++;
             }
+            // Indicated end of mDNS browsing
```
Comment (bzbarsky-apple) [inline src/platform/ESP32/DnssdImpl.cpp:314]:
```suggestion
            // Indicate end of mDNS browsing
```

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 9]
```diff
@@ -2546,18 +2545,16 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnDnsBrowseResult(otEr
 
         VerifyOrExit(error == CHIP_NO_ERROR, );
 
-        DnsResult * dnsResult = Platform::New<DnsResult>(aContext, MapOpenThreadError(aError));
+        DnsResult * dnsResult = Platform::New<DnsResult>(aContext, CHIP_NO_ERROR);
```
Comment (bzbarsky-apple) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2548]: Why this change?

File: src/lib/dnssd/platform/Dnssd.h [hunk 10]
```diff
@@ -102,10 +102,15 @@ using DnssdResolveCallback = void (*)(void * context, DnssdService * result, con
...
+ * It is possible that this callback will be called multiple number of times.
+ * Only in case when the caller passes the CHIP_END_OF_INPUT as an error code
+ * we might assume that the mDNS browsing has ended. In that case the callback
+ * function can safely release its resources.
+ *
  * @param[in] context       The context passed to ChipDnssdBrowse or ChipDnssdResolve.
  * @param[in] services      The service list, can be nullptr.
  * @param[in] servicesSize  The size of the service list.
- * @param[in] error         The error code.
+ * @param[in] error         The error code or CHIP_END_OF_INPUT.
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/platform/Dnssd.h:113]: So what's actually going on here is that any error code is treated as "last call" (that needs to be clearly documented), but CHIP_END_OF_INPUT means "last call, but process the data too"?

If so, that all needs to be clearly documented.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp (hunk 9 above)
Comment (Arkq) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2548]: At this line number there is no way that the `aError` indicates an error. It is verified on line 2530. I though that it will easier to read the code if there will be no error conversion but simply `CHIP_NO_ERROR` to satisfy the `DnsResult` API - it's easier to follow that `DispatchBrowse` is called with success result code.

Comment (bzbarsky-apple) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2548]: >  It is verified on line 2530

Ah, perfect.  That was outside the diff context.  :)

File: src/lib/dnssd/platform/Dnssd.h (hunk 10 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h]: I've updated the docstring. I hope that right now it will be more clear which error code is responsible for stopping browsing and which process data.

File: src/lib/dnssd/platform/Dnssd.h (hunk 7 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h:239]: Yes, I'm planing to do a follow up for that, but to be honest I've faced some issues. I'm not sure right now whether start/stop approach is the easiest way of achieving the goal that I've set for myself - I'd like that Tizen will be able to discover more than one Matter device. I'm still doing some research what approach would be better or how to apply my current proposal, so please do not close this conversation for now.

Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h]: I've dropped this new ChipDnssdStopBrowse API in favor of #21553

File: src/lib/dnssd/platform/Dnssd.h (hunk 1 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h]: I've dropped this new ChipDnssdStopBrowse API in favor of https://github.com/project-chip/connectedhomeip/pull/21553

**Comment by woody-apple** [comment]
This has had more than enough time for review, and all comments are resolved, fast tracking.

**Comment by woody-apple** [comment]
@Arkq can you take a peek at the conflict here please?

**Review by woody-apple — APPROVED** [review]

File: src/lib/dnssd/Discovery_ImplPlatform.cpp [hunk 11]
```diff
@@ -151,6 +151,12 @@ static void HandleNodeBrowse(void * context, DnssdService * services, size_t ser
 {
     ResolverDelegateProxy * proxy = static_cast<ResolverDelegateProxy *>(context);
 
+    if (error != CHIP_NO_ERROR && error != CHIP_END_OF_INPUT)
```
Comment (andy31415) [inline src/lib/dnssd/Discovery_ImplPlatform.cpp:154]: END_OF_INPUT does not seem to be an error. Could we design our logic in some other way so that `error` actually is used as an error?

File: src/lib/dnssd/platform/Dnssd.h [hunk 12]
```diff
@@ -102,10 +102,18 @@ using DnssdResolveCallback = void (*)(void * context, DnssdService * result, con
  * The callback function SHALL NOT take the ownership of the service pointer or
  * any pointer inside this structure.
  *
+ * It is possible that this callback will be called multiple times. In case when
```
Comment (andy31415) [inline src/lib/dnssd/platform/Dnssd.h:105]: I find this awkward/problematic to maintain. I expect errors to actually mean errors. This seems to be some form of ternary of in one of `No error, more callbacks expected`, `no error, final callback`, `has error, final callback`.

I would prefer some better split of arguments and their meaning. Having `CHIP_END_OF_INPUT` be used as an error code that means success is not ideal.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Please make the logic of error codes and final/non-final callback better delimited.

Re-using the error callback seems to make it more likely that not all code paths would be updated to handle the newly introduced and used error code.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 13]
```diff
@@ -2570,18 +2569,16 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnDnsBrowseResult(otEr
...
```
Comment (andy31415) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2572]: previous code was doing a map of aError. New code just uses CHIP_NO_ERROR. was the old code a bug? 

Same question for updates below.

Comment (andy31415) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2572]: I now see above `VerifyOrExit(aError == OT_ERROR_NONE, error = MapOpenThreadError(aError));` ... so old code seems indeed awkward.

**Comment by Arkq** [comment]
> Please make the logic of error codes and final/non-final callback better delimited.

OK, I'll try to redesign this. However, this will have to wait some time - I'll be off the computer (🌞🌴) until the end of this month.
Also, I will trying to reproduce scenario which will show that the previous code for OpenThread platform was indeed buggy.

File: src/lib/dnssd/platform/Dnssd.h (hunk 12 above)
Comment (Arkq) [reply src/lib/dnssd/platform/Dnssd.h]: I've added new parameter for the callback function, so right now the "end-of-input" will be indicated in an explicit way.
